# Q1. Create an MLP model with 16 hidden layers using "mnist_784" dataset from sklearn and improve the result using hyperparameter tuning. Hints.

In [1]:
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

In [2]:
# Load the MNIST dataset
mnist = fetch_openml("mnist_784")

C:\Users\Arun\anaconda3\lib\site-packages\sklearn\datasets\_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [5]:
print(mnist.keys())

dict_keys(['data', 'target', 'frame', 'categories', 'feature_names', 'target_names', 'DESCR', 'details', 'url'])


In [6]:
data = mnist['data']
target = mnist['target']

In [7]:
data.head()

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   pixel781  pixel782  pixel783  pixel784  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 784 columns]

In [9]:
target.head()

0    5
1    0
2    4
3    1
4    9
Name: class, dtype: category
Categories (10, object): ['0', '1', '2', '3', ..., '6', '7', '8', '9']

In [11]:

X = mnist.data.astype(float)
y = mnist.target.astype(int)

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create an MLP model with 16 hidden layers
mlp = MLPClassifier(hidden_layer_sizes=(16,) * 16, max_iter=100)

# Train the model
mlp.fit(X_train, y_train)

# Predict on the test set
y_pred = mlp.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy without hyperparameter tuning:", accuracy)

Accuracy without hyperparameter tuning: 0.9478571428571428


C:\Users\Arun\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


In [12]:
# Hyperparameter tuning using GridSearchCV
param_grid = {
    'hidden_layer_sizes': [(16,) * 16],
    'activation': ['relu', 'tanh'],
    'solver': ['adam', 'sgd'],
    'learning_rate': ['constant', 'adaptive'],
}

grid_search = GridSearchCV(MLPClassifier(max_iter=100), param_grid, cv=3, n_jobs=-1)
grid_search.fit(X_train, y_train)

best_mlp = grid_search.best_estimator_

# Predict on the test set using the best model
y_pred_tuned = best_mlp.predict(X_test)

# Calculate accuracy with tuned model
accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
print("Accuracy with hyperparameter tuning:", accuracy_tuned)

Accuracy with hyperparameter tuning: 0.9481428571428572


C:\Users\Arun\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


In [13]:
print("Best hyperparameters:", grid_search.best_params_)

Best hyperparameters: {'activation': 'relu', 'hidden_layer_sizes': (16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16), 'learning_rate': 'constant', 'solver': 'adam'}


**Below is for 1000 iterations.**

In [14]:
# Create an MLP model with 16 hidden layers and set max_iter to a reasonable value
mlp = MLPClassifier(hidden_layer_sizes=(16,) * 16, max_iter=1000)  # You can adjust the value as needed

# Train the model
mlp.fit(X_train, y_train)

# Predict on the test set
y_pred = mlp.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy without hyperparameter tuning:", accuracy)

Accuracy without hyperparameter tuning: 0.9416428571428571


# Q2.  Explain your findings and report best hyperparameter values.

**Initial Model Without Hyperparameter Tuning:**  

The initial MLP model with 16 hidden layers and max_iter=100 achieved an accuracy of approximately 0.9479 on the test set.  

**Model with Best Hyperparameters:**  

A new MLP model was trained using the best hyperparameters obtained from the grid search.  
The new model achieved an accuracy of approximately 0.9481 on the test set, which is slightly better than the initial model's accuracy.  

**Conclusion:**  

We found slightly better results using best hyperparameters but the computation cost and time used in the process was a lot. These slight differences can be very crucial for some cases.  
Also if not the case then we must see if this cost is reasonable for the objective we want to achieve.

## Reference

Deyuan Deng, A. (2019, February 10). Machine Learning Notes. GitHub. https://github.com/amaliujia/ml-notes